discord/ext/commands/core.pyi

```diff
@@ -0,0 +1,161 @@
+from inspect import Parameter
+
+from .context import Context
+from .cooldowns import CooldownMapping, BucketType, Cooldown
+from .cog import Cog
+from ._types import _BaseCommand
+
+from typing import Any, Optional, Union, Callable, Dict, Iterator, Coroutine, Type, ValuesView, List, TypeVar, Mapping, Generic, overload
+from typing_extensions import Protocol
+
+_CoroType = Callable[..., Coroutine[Any, Any, Any]]
+
+_CT = TypeVar('_CT', bound=Context)
+_C = TypeVar('_C', bound=_CoroType)
+_CMD = TypeVar('_CMD', bound=Command[Any])
+_F = TypeVar('_F', bound=Union[_CoroType, Command[Any]])
+_T_contra = TypeVar('_T_contra', contravariant=True)
+
+class _CheckPredicate(Protocol[_T_contra]):
+    def __call__(self, __ctx: _T_contra) -> Union[bool, Coroutine[Any, Any, bool]]: ...
+
+class _CheckDecorator(Protocol):
+    def __call__(self, __func: _F) -> _F: ...
+
+class _InvokeCallback(Protocol[_T_contra]):
+    async def __call__(self, __ctx: _T_contra) -> None: ...
+
+class Command(_BaseCommand, Generic[_CT]):
+    name: str
+    callback: _CoroType
+    help: str
+    brief: str
+    usage: str
+    aliases: List[str]
+    enabled: bool
+    parent: Optional[Command[_CT]]
+    checks: List[_CheckPredicate[_CT]]
+    description: str
+    hidden: bool
+    rest_is_raw: bool
+    ignore_extra: bool
+    cooldown_after_parsing: bool
+    params: Mapping[str, Parameter]
+    _buckets: CooldownMapping
+    cog: Optional[Cog[_CT]]
+
+    def __init__(self, func: _CoroType, *, name: str = ..., enabled: bool = ...,
+                 help: Optional[str] = ..., brief: Optional[str] = ..., usage: Optional[str] = ...,
+                 aliases: List[str] = ..., description: str = ..., hidden: bool = ...,
+                 rest_is_raw: bool = ..., ignore_extra: bool = ..., cooldown_after_parsing: bool = ...,
+                 checks: List[_CheckPredicate[_CT]] = ..., cooldown: Cooldown = ..., parent: _BaseCommand = ...,
+                 cog: Optional[Cog[_CT]] = ...) -> None: ...
```

Comment:
I think all of these kwargs need to be in `Optional`—it's possible to create a `Command` with just `func`.

Reply:
`Optional` is for things that can be None. The reason you can create a `Command` with just `func` is that these arguments all have defaults, which is reflected in these type annotations.

Reply:
Oh, you're right. I confused defaults with optionals.